types/block.go

```diff
@@ -434,65 +436,130 @@ func (h *Header) StringIndented(indent string) string {
+type CommitSig struct {
+	BlockIDFlag      BlockIDFlag `json:"block_id_flag"`
+	ValidatorAddress Address     `json:"validator_address"`
+	Timestamp        time.Time   `json:"timestamp"`
+	Signature        []byte      `json:"signature"`
+}
+
+// NewCommitSigForBlock returns new CommitSig with BlockIDFlagCommit.
+func NewCommitSigForBlock(signature []byte, valAddr Address, ts time.Time) *CommitSig {
+	return &CommitSig{
+		BlockIDFlag:      BlockIDFlagCommit,
+		ValidatorAddress: valAddr,
+		Timestamp:        ts,
+		Signature:        signature,
+	}
 }
 
-// toVote converts the CommitSig to a vote.
-// TODO: deprecate for #1648. Converting to Vote will require
-// access to ValidatorSet.
-func (cs *CommitSig) toVote() *Vote {
+func (cs *CommitSig) String() string {
 	if cs == nil {
-		return nil
+		return nilCommitSigStr
+	}
+
+	return fmt.Sprintf("CommitSig{%X by %X on %v @ %s}",
+		cmn.Fingerprint(cs.Signature),
+		cmn.Fingerprint(cs.ValidatorAddress),
+		cs.BlockIDFlag,
+		CanonicalTime(cs.Timestamp))
+}
+
+// BlockID returns the Commit's BlockID if CommitSig indicates signing,
+// otherwise - empty BlockID.
+func (cs *CommitSig) BlockID(commitBlockID BlockID) BlockID {
+	var blockID BlockID
+	switch cs.BlockIDFlag {
+	case BlockIDFlagAbsent:
+		blockID = BlockID{}
+	case BlockIDFlagCommit:
+		blockID = commitBlockID
+	case BlockIDFlagNil:
+		blockID = BlockID{}
+	default:
+		panic(fmt.Sprintf("Unknown BlockIDFlag: %v", cs.BlockIDFlag))
 	}
-	v := Vote(*cs)
-	return &v
+	return blockID
+}
+
+// ValidateBasic performs basic validation.
+func (cs *CommitSig) ValidateBasic() error {
+	switch cs.BlockIDFlag {
+	case BlockIDFlagAbsent:
+	case BlockIDFlagCommit:
+	case BlockIDFlagNil:
+	default:
+		return fmt.Errorf("unknown BlockIDFlag: %v", cs.BlockIDFlag)
+	}
+
+	if len(cs.ValidatorAddress) != crypto.AddressSize {
+		return fmt.Errorf("expected ValidatorAddress size to be %d bytes, got %d bytes",
+			crypto.AddressSize,
+			len(cs.ValidatorAddress),
+		)
+	}
+
+	// NOTE: Timestamp validation is subtle and handled elsewhere.
+
+	if len(cs.Signature) == 0 {
+		return errors.New("signature is missing")
+	}
+	if len(cs.Signature) > MaxSignatureSize {
+		return fmt.Errorf("signature is too big (max: %d)", MaxSignatureSize)
+	}
+
+	return nil
 }
 
 //-------------------------------------
 
 // Commit contains the evidence that a block was committed by a set of validators.
 // NOTE: Commit is empty for height 1, but never nil.
 type Commit struct {
-	// NOTE: The Precommits are in order of address to preserve the bonded ValidatorSet order.
-	// Any peer with a block can gossip precommits by index with a peer without recalculating the
-	// active ValidatorSet.
+	// NOTE: The Precommits are in order of address to preserve the bonded
+	// ValidatorSet order.
+	// Any peer with a block can gossip precommits by index with a peer without
+	// recalculating the active ValidatorSet.
+	Height     int64        `json:"height"`
```

Comment:
Just for my own understanding, can you explain why we need to add `Height` and `Round` to `Commit`?

Reply:
this adr may help explain things more in depth: https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-025-commit.md

Reply:
They're in the individual votes, and we need both because we need a commit to come from a particular round of consensus, we don't want a commit with some votes from one round and some from another. But since they are in every vote, we move them into the Commit instead to deduplicate.